plutus-core/plutus-core/test/Value/Spec.hs

```diff
@@ -191,32 +257,79 @@ checkBookkeeping v =
   actualMaxInnerSize = V.maxInnerSize v
   expectedSize = sum $ Map.map Map.size (V.unpack v)
   actualSize = V.totalSize v
-  expectedNeg = length [amt | inner <- Map.elems (V.unpack v), amt <- Map.elems inner, amt < 0]
+  expectedNeg =
+    length [q | inner <- Map.elems (V.unpack v), q <- Map.elems inner, V.unQuantity q < 0]
   actualNeg = V.negativeAmounts v
 
 checkInvariants :: Value -> Property
 checkInvariants (V.unpack -> v) =
   property ((not . any Map.null) v)
-    .&&. property ((not . any (elem 0)) v)
+    .&&. property ((not . any (elem V.zeroQuantity)) v)
 
-prop_unValueDataValidatesCurrency :: ValueAmount -> Property
-prop_unValueDataValidatesCurrency (ValueAmount amt) =
+prop_unValueDataValidatesCurrency :: V.Quantity -> Property
+prop_unValueDataValidatesCurrency quantity =
   forAll gen33Bytes $ \c ->
     forAll gen32BytesOrFewer $ \t ->
-      let d = Map [(B c, Map [(B t, I amt)])]
+      let d = Map [(B c, Map [(B t, I (V.unQuantity quantity))])]
        in case V.unValueData d of
             BuiltinFailure{} -> property True
             _                -> property False
 
-prop_unValueDataValidatesToken :: ValueAmount -> Property
-prop_unValueDataValidatesToken (ValueAmount amt) =
+prop_unValueDataValidatesToken :: V.Quantity -> Property
+prop_unValueDataValidatesToken quantity =
   forAll gen32BytesOrFewer $ \c ->
     forAll gen33Bytes $ \t ->
-      let d = Map [(B c, Map [(B t, I amt)])]
+      let d = Map [(B c, Map [(B t, I (V.unQuantity quantity))])]
+       in case V.unValueData d of
+            BuiltinFailure{} -> property True
+            _                -> property False
+
+prop_unValueDataValidatesQuantityMin :: Property
+prop_unValueDataValidatesQuantityMin =
+  forAll gen32BytesOrFewer $ \c ->
+    forAll gen32BytesOrFewer $ \t ->
+      let amt = V.minQuantity - 1
```

Comment:
Similarly, you can generate more quantities here. You can also generate a Value with multiple quantities (perhaps 90% valid and 10% invalid), and then check that `unValueData` succeeds if and only if all quantities are valid.